packages/api/tests/lambdas/test-verifyProviderSecretsMigration.js

```diff
@@ -0,0 +1,150 @@
+
+test.after.always(() => S3.recursivelyDeleteS3Bucket(process.env.system_bucket));
+
+test.serial('verifyProviderSecretsMigration passes if empty credentials are found', async (t) => {
+  const provider = fakeProviderFactory({
+    protocol: 'ftp',
+    encrypted: false
+  });
+
+  delete provider.username;
+  delete provider.password;
+
+  await dynamodbDocClient().put({
+    TableName: process.env.ProvidersTable,
+    Item: { ...provider, createdAt: Date.now() }
+  }).promise();
+
+  await t.notThrowsAsync(handler());
+});
+
+test.serial('verifyProviderSecretsMigration passes if KMS credentials are found', async (t) => {
+  const { providerModel } = t.context;
+
+  const provider = fakeProviderFactory({
+    protocol: 'ftp',
+    username: 'my-username',
+    password: 'my-password'
+  });
+
+  await providerModel.create(provider);
+
+  await t.notThrowsAsync(handler());
+});
+
+test.serial('verifyProviderSecretsMigration fails if plaintext credentials are found', async (t) => {
+  const provider = fakeProviderFactory({
+    protocol: 'ftp',
+    encrypted: false,
+    username: 'my-username',
+    password: 'my-password'
+  });
+
+  await dynamodbDocClient().put({
+    TableName: process.env.ProvidersTable,
+    Item: { ...provider, createdAt: Date.now() }
+  }).promise();
+
+  const err = await t.throwsAsync(handler());
+
+  t.is(
+    err.message,
+    `Provider ${provider.id} has plaintext username or password. Must invoke the providerSecretsMigration Lambda function.`
+  );
+});
+
+test.serial('verifyProviderSecretsMigration fails if S3 keypair credentials are found', async (t) => {
+  const provider = fakeProviderFactory({
+    protocol: 'ftp',
+    encrypted: true,
+    username: await S3KeyPairProvider.encrypt('my-username'),
+    password: await S3KeyPairProvider.encrypt('my-password')
+  });
+
+  await dynamodbDocClient().put({
+    TableName: process.env.ProvidersTable,
+    Item: { ...provider, createdAt: Date.now() }
+  }).promise();
+
+  const err = await t.throwsAsync(handler());
+
+  t.is(
+    err.message,
+    `Provider ${provider.id} credentials could not be decrypted using KMS. Must invoke the providerSecretsMigration Lambda function.`
+  );
+});
+
+test.serial('verifyProviderSecretsMigration verifies all providers', async (t) => {
+  const { providerModel } = t.context;
+
+  const kmsProvider = fakeProviderFactory({
+    protocol: 'ftp',
+    username: 'my-username',
+    password: 'my-password'
+  });
+
+  await providerModel.create(kmsProvider);
+
+  const s3Provider = fakeProviderFactory({
+    protocol: 'ftp',
+    encrypted: true,
+    username: await S3KeyPairProvider.encrypt('my-username'),
+    password: await S3KeyPairProvider.encrypt('my-password')
+  });
+
+  await dynamodbDocClient().put({
+    TableName: process.env.ProvidersTable,
+    Item: { ...s3Provider, createdAt: Date.now() }
+  }).promise();
+
+  const err = await t.throwsAsync(handler());
+
+  t.is(
```

Comment:
So this test works because you're only creating one provider that has KMS encryption, which will not throw an error, and the one using S3 key pair encryption that will throw an error. If you had a test where you setting up two failing cases, you wouldn't be able to guarantee which one throws an error first. 

I'm assuming you know this and are fine with the lambda working this way, but the test says `verifies all providers`, but with the Lambda as written it will stop verification as soon as it hits the first failing case

Reply:
As soon as we find a single provider that's not encrypted using KMS, we know that the migration has not been run, so we stop everything right there.

Reply:
So by `verifies all providers`, we mean `throws error if any provider is not encrypted using KMS`. I'm not suggesting that you rename the test, just clarifying

Reply:
Correct.